File: src/main/GO/leetcode/lists/offer/Offer24.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2022/12/29 18:36
 */

func reverseList(head *ListNode) *ListNode {
	var dfs func(node *ListNode) *ListNode
	dfs = func(node *ListNode) *ListNode {
		if node == nil || node.Next == nil {
			return node
		}
		next := dfs(node.Next)
		node.Next.Next = node
		node.Next = nil
		return next
	}
	return dfs(head)
}

File: java/src/leetcode/offer/Offer24.java
package leetcode.offer;

import common.ListNode;

/**
 * @author jingxinwu
 * @date 2021-11-21 6:11 下午
 */
public class Offer24 {

    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode pre = null;
        ListNode cur = head;
        while (cur != null) {
            ListNode temp = cur.next;
            cur.next = pre;
            pre.next = temp;

            pre = cur;
            cur = temp;
        }

        return pre;
    }

}
